packages/xstate-store/src/store.ts

```diff
@@ -187,10 +187,7 @@ export function createStoreWithProducer<
   TContext extends StoreContext,
   TEventPayloadMap extends EventPayloadMap
 >(
-  producer: (
-    context: TContext,
-    recipe: (context: TContext) => void
-  ) => TContext,
+  producer: <T>(context: T, recipe: (context: T) => void) => T,
```

Comment:
This feels like a more straightforward thing to do here if you want to avoid complex generic producers from affecting ur inference in weird ways:
```suggestion
  producer: NoInfer<(
    context: TContext,
    recipe: (context: TContext) => void
  ) => TContext>,
```

Reply:
Well, the producer shouldn't care about the `TContext` type anyway - it's a generic producer.

Reply:
You don't want to require a generic producer - you just happen to receive one here (the `immer`'s producer is generic).

Take a look at what worked before your change: [TS playground](https://www.typescriptlang.org/play/?ts=5.4.5#code/C4TwDgpgBAysD2AnCBheA7YEAewoF4oAlCAYyQBMAeAZ2EQEt0BzAGigEN0QA+AbgBQAimQA2HZFABmAV3SlgDDFFLIOWOEggB1BsAAWABUTwKM0hERUAKmkw48DiOgo1YCZHay4eACgFQUGAmZhaIAFxQ-oGB5Pa4kbYY3sCsATHIpAyQkb5xKYleDgCUBDxQAG7wDBRpgaX45UnxqelMegwcokUJUM0pacWRAN7pzBDAMOgcYDT68MC5DeXDKskOheu4UAC+gntCcXRQNAwAtmCiEMam5tCEecDYI2tyi1DoMmcARpa7yyonoIBEc8HQtAQVGoNB4dHojCE7ohfKcLlcbqEIOxRrF4G9IgAGNI7YrA8HIAB040m01m82AvmKFPyDhO6gYNCkDAgblW5HxHy+v0Qu0EQA). Requiring a generic producer breaks this: [TS playground](https://www.typescriptlang.org/play/?ts=5.4.5#code/C4TwDgpgBAysD2AnCBheA7YEAewoF4oAlCAYyQBMAeAZ2EQEt0BzAGigEN0QA+AbgBQAimQA2HZFABmAV3SlgDDFFLIOWOEggB1BsAAWABUTwKM0hERUAKmkw48DiOgo1YCZHay4eACgFQUGAmZhaIAFxQNn7k9riR1uzIpAyQkb6x3sAJAJQEPFAAbvAMFHn4BYkBUEx6DByiXg4JTbisAjmRAN7VzBDAMOgcYDT68NlQvuUFXSoYWS3zDlAAvoJrQrF0UDQMALZgohDGpubQhBnA2N1zchPoMnsARpar0ypXggJbeHRaBCo1BoPDo9EYQmdEL5dgcjidQhB2D1AuQ7pEAAztFY5L5-ZAAOj6AyGIzGwCm+Myyxo6gYNCkDAgblmqMwkQez1eGyAA). With `NoInfer` both cases work OK.